File: src/main/java/eu/codetopic/utils/NetworkManager.java
/*
 * Copyright 2017 Jiří Kuchyňka (Anty)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package eu.codetopic.utils;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.PowerManager;

public final class NetworkManager {

    private static final String LOG_TAG = "NetworkManager";

    private static Context mContext;

    private NetworkManager() {
    }

    public static void init(Context context) {
        if (mContext != null) throw new IllegalStateException(LOG_TAG + " is still initialized");
        mContext = context.getApplicationContext();
    }

    public static NetworkInfo getInfo() {
        ConnectivityManager manager = (ConnectivityManager) mContext.getSystemService(Context.CONNECTIVITY_SERVICE);
        return manager.getActiveNetworkInfo();
    }

    public static boolean isConnected(NetworkType type) {
        if (Build.VERSION.SDK_INT >= 23) {
            PowerManager powerManager = (PowerManager) mContext.getSystemService(Context.POWER_SERVICE);
            if (powerManager.isDeviceIdleMode()) return false;
        }

        NetworkInfo info = getInfo();
        if (info != null && info.isConnected()) {
            if (type == NetworkType.ANY) {
                return true;
            } else if (type == NetworkType.MOBILE) {
                return info.getType() == ConnectivityManager.TYPE_MOBILE;
            } else if (type == NetworkType.WIFI) {
                return info.getType() == ConnectivityManager.TYPE_WIFI;
            }
        }
        return false;
    }

    public enum NetworkType {
        ANY, WIFI, MOBILE
    }
}

File: src/main/java/eu/codetopic/utils/network/NetworkManager.kt
/*
 * utils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.utils.network

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.os.Build
import android.os.PowerManager

object NetworkManager {

    private const val LOG_TAG = "NetworkManager"

    private lateinit var connManager: ConnectivityManager
    private lateinit var powerManager: PowerManager

    @JvmStatic
    val info: NetworkInfo? get() = connManager.activeNetworkInfo

    fun init(context: Context) {
        if (::connManager.isInitialized)
            throw IllegalStateException("$LOG_TAG is still initialized")
        context.applicationContext.run {
            connManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            powerManager = getSystemService(Context.POWER_SERVICE) as PowerManager
        }
    }

    @JvmStatic
    fun isConnected(type: NetworkType): Boolean {
        if (Build.VERSION.SDK_INT >= 23 && powerManager.isDeviceIdleMode) return false

        return info?.takeIf { it.isConnected }?.let {
            when (type) {
                NetworkType.ANY -> true
                NetworkType.MOBILE -> it.type == ConnectivityManager.TYPE_MOBILE
                NetworkType.WIFI -> it.type == ConnectivityManager.TYPE_WIFI
            }
        } ?: false
    }

    enum class NetworkType {
        ANY, WIFI, MOBILE
    }
}
